src/timesync/nts_tls.c

```diff
@@ -0,0 +1,247 @@
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <assert.h>
+
+#ifdef USE_GNUTLS
+#include <gnutls/gnutls.h>
+#else
+#include <openssl/ssl.h>
+#endif
+
+#include "nts.h"
+
+int NTS_TLS_extract_keys(
+                NTS_TLS *opaque,
+                NTS_AEADAlgorithmType aead,
+                uint8_t *c2s,
+                uint8_t *s2c,
+                int key_capacity) {
+
+#ifdef USE_GNUTLS
+        gnutls_session_t session = (void *)opaque;
+#else
+        SSL *session = (void *)opaque;
+#endif
+
+        uint8_t *keys[] = { c2s, s2c };
+        const char label[30] = { "EXPORTER-network-time-security" }; /* note: this does not include the zero byte */
+
+        const struct NTS_AEADParam *info = NTS_get_param(aead);
+        if (!info)
+                return -3;
+        else if (info->key_size > key_capacity)
+                return -2;
+
+        for (int i=0; i < 2; i++) {
+                const char context[5] = { 0, 0, (aead >> 8) & 0xFF, aead & 0xFF, i };
```

Comment:
this is really weird. you use "char" here, which is signed on most systems. but you then assign some value & 0xFF to it, which suggests you mean the 0…255 range. thus you are always triggering an overflow here. use uint8_t hence